### PR TITLE
[5.4] Higher-order messages for the collection class

### PR DESCRIPTION
**Pull request by franzliedke**
Howdy - long time, no see! :)

I've recently come across [an article](http://natpryce.com/articles/000535.html) about a concept called Higher-Order Messages. That particular article was about Ruby, but it intrigued me to implement this for Laravel. And it turns out it's very easy!

This lets you turn some very common collection use-cases into concise one-liners.

`filter` example
~~~php
// Instead of this:
$collection->filter(function($item) {
  return $item->field;
});

// Write this:
$collection->where->field;
~~~

It also works for other collection methods:
~~~php
// Instead of this:
$collection->each(function($item) {
  return $item->printYourself();
});

// Write this:
$collection->do->printYourself();
~~~

---

**This is just a proposal. If you're interested in merging this into Laravel, I'll gladly provide the unit tests and clean up the code. If not, I have saved some time and will probably publish a package.**

Naming: I used the conventions from the aforementioned article, but this probably requires some input to make it consistent with the rest of the collection.

### CODE REVIEW

File: src/Illuminate/Support/Collection.php [hunk 1]
```diff
@@ -1350,3 +1369,36 @@ protected function getArrayableItems($items)
...
+    {
+        return $this->collection->{$this->method}(function($value) use ($name, $arguments) {
+            return call_user_func_array([$value, $name], $arguments);
+        });
+    }
+
+    public function __get($name)
+    {
+        return $this->collection->{$this->method}(function($value) use ($name) {
+            return $value->$name;
```
Comment (CristianLlanos) [inline src/Illuminate/Support/Collection.php]: Perhaps we shouldn't assume the value will be an object. We also use collections for arrays.

**Review by CristianLlanos — COMMENTED** [review]
It looks interesting :)

**Comment by taylorotwell** [comment]
That's some clever stuff! 😈

**Comment by franzliedke** [comment]
Woah, this wasn't ready yet. I wanted to add tests, and docs, and an `offsetGet` method.

File: src/Illuminate/Support/Collection.php [hunk 2]
```diff
@@ -1266,6 +1266,25 @@ public function toBase()
...
+            'where' => 'filter',
+            'unless' => 'reject',
+            'do' => 'each',
+            'extract' => 'map',
+            'sum' => 'sum',
+            'inOrderOf' => 'sortBy',
+            'inReverseOrderOf' => 'sortByDesc',
+        ];
+
+        if (isset($nameToMethod[$name])) {
```
Comment (franzliedke) [inline src/Illuminate/Support/Collection.php]: Damn, that's what I get for refactoring this method at the last moment.

File: src/Illuminate/Support/Collection.php (hunk 1 above)
Comment (alexbowers) [inline src/Illuminate/Support/Collection.php]: @franzliedke I fixed this in a separate PR to allow getting by object or array

File: src/Illuminate/Support/Collection.php [hunk 3]
```diff
@@ -1266,6 +1266,25 @@ public function toBase()
...
+    {
+        $nameToMethod = [
+            'where' => 'filter',
+            'unless' => 'reject',
+            'do' => 'each',
+            'extract' => 'map',
+            'sum' => 'sum',
+            'inOrderOf' => 'sortBy',
+            'inReverseOrderOf' => 'sortByDesc',
+        ];
```
Comment (mathieutu) [inline src/Illuminate/Support/Collection.php]: With that array we can't make the `$invoices->each->pay();` announced in the [LN article ](https://laravel-news.com/2016/11/higher-order-messaging). Can we ? We have to do `$invoices->do->pay();` instead. Maybe am I missing something somewhere else ? In this case, I think we should use the original methods too.
Thx.

Comment (franzliedke) [inline src/Illuminate/Support/Collection.php]: That was just the original implementation with the method names from the article that I linked in the original description. I didn't consider this finished, but Taylor merged it and changed it to what was explained in the Laravel News article.

Comment (mathieutu) [inline src/Illuminate/Support/Collection.php]: Yeah I've seen that afterwards. Too bad I liked the aliases you made...
But I've seen too that the merge and the update were fast and uncontrolled ! ^^"

Comment (envision) [inline src/Illuminate/Support/Collection.php]: Personally I would prefer $invoices->each()->pay(); format over $invoices->each->pay();

**Pull request merged**